### ISSUE TITLE
#upstream_remote and deploy via copy

### ISSUE THREAD

**Issue by dnd** [jrochkind/cap_git_tools#2]
I'm having a problem with git:guard_upstream, or more specifically `#upstream_remote`. Because of server accessibility, I have to do my deployments on our network using deploy_via copy. Doing so the repository is set to `.`. This causes `upstream_remote` to fail and return the upstream name as `.`, which is obviously incorrect. 

I have forked the project, and plan on fixing to make it work for me, but I wanted to see if you had an idea/preference on how to make it work? Either a configurable variable, or detecting the it's a copy deployment and if there is only one remote listed use that?

**Comment by jrochkind** [comment]
Hmm, can you simply wire up your recipe to not include the `git:guard_upstream` task in that scenario?  

Like, maybe, instead of just including `before "deploy:update_code", "git:guard_upstream"` in your recipe, something like:

```
unless fetch(:no_guard_upstream) == "true"
   before "deploy:update_code",  "git:guard_upstream"
end
```

Or use whatever other condition is appropriate for your scenario, instead of that proposed custom cap variable `no_guard_upstream`. I _think_ you can wrap the cap `before` in a condition of your choosing like that?

This is why cap_git_tools doesn't wire up any tasks for you but lets you wire them up yourself -- so you can do so however meets your need, using whatever logic you need. 

If there's a simple way to make it work with just plain ruby code written locally that exactly meets your needs, that seems preferable to me to any change to cap_git_tools. But if there's no convenient way to do that, and you can make it so there is with a pull request, I'm happy to consider one. 

Thanks for checking in. And thanks for letting me know you're using cap_git_tools, it wasn't entirely clear to me whether anyone other than me was using it!

**Comment by jrochkind** [comment]
(Oh, and I think I understand that you want in some scenarios guard_upstream to _not run_, because it can't be run, right?

If instead you want guard_upstream to somehow be run _differently_ in some scenarios, rather than be suppressed -- then sorry, I misunderstood, and would have a different answer. :) )

**Comment by dnd** [comment]
I still want the code to be guarded though. The local repo that is being deployed from should still be completely up to date in both directions before the deployment code copies it to the remote server. Otherwise it could still potentially deploy outdated code.

**Comment by jrochkind** [comment]
Ah, yeah, sorry, I misunderstood. 

Okay. Yeah, I don't really understand the best way to do this -- I have to admit I'm not a git expert, and basically copy and pasted the git interaction that's in there from other gems that did it. 

So I don't have much guidance to give. Happy to see a pull request, if you think you know how to do it properly. 

My preference would be for whichever way leaves the cap_git_tools code looking simpler and easier to understand compared to the alternatives -- if there's any difference. I guess I'd assume it would be better to make the code more robust and 'just work' in more situations, vs adding another configuration variable.  Unless the code to make it 'just work' ends up complicated and making other assumptions that might be wrong and give people more trouble down the line. So, yeah, basically, I'm not really sure, but might have more ideas after seeing the pull request. heh. 

Thanks!

(And sadly, cap_git_tools has no tests, yeah).

**Comment by jrochkind** [comment]
See #3
